src/kmgmt/ec.c

```diff
@@ -0,0 +1,924 @@
+}
+
+static void *p11prov_ec_gen_init(void *provctx, int selection,
+                                 const OSSL_PARAM params[])
+{
+    struct key_generator *ctx = NULL;
+    int ret;
+
+    P11PROV_debug("ec gen_init %p", provctx);
+
+    /* we need to allow to initialize a generation of just domain parameters,
+     * as this is used by OpenSSL for ECDH, to set the expected parameters
+     * first and then import the received public peer key */
+    if ((selection & OSSL_KEYMGMT_SELECT_ALL) == 0) {
+        P11PROV_raise(provctx, CKR_ARGUMENTS_BAD, "Unsupported selection");
+        return NULL;
+    }
+
+    ctx = p11prov_kmgmt_gen_init(provctx, CKK_EC, CKM_EC_KEY_PAIR_GEN);
+    if (!ctx) {
+        return NULL;
+    }
+
+    /* set defaults */
+    ctx->data.ec.ec_params = prime256v1_param;
+    ctx->data.ec.ec_params_size = sizeof(prime256v1_param);
+
+    /* is this a parameter generator request ? */
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0) {
+        ctx->data.ec.paramgen = true;
+    }
+
+    ret = p11prov_ec_gen_set_params(ctx, params);
+    if (ret != RET_OSSL_OK) {
+        p11prov_kmgmt_gen_cleanup(ctx);
+        ctx = NULL;
+    }
+    return ctx;
+}
+
+static const OSSL_PARAM *p11prov_ec_gen_settable_params(void *genctx,
+                                                        void *provctx)
+{
+    static OSSL_PARAM p11prov_ec_params[] = {
+        OSSL_PARAM_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME, NULL, 0),
+        OSSL_PARAM_utf8_string(P11PROV_PARAM_URI, NULL, 0),
+        OSSL_PARAM_utf8_string(P11PROV_PARAM_KEY_USAGE, NULL, 0),
+        OSSL_PARAM_END,
+    };
+    return p11prov_ec_params;
+}
+
+static int p11prov_ec_gen_set_params(void *genctx, const OSSL_PARAM params[])
+{
+    struct key_generator *ctx = (struct key_generator *)genctx;
+    const OSSL_PARAM *p;
+
+    if (!ctx) {
+        return RET_OSSL_ERR;
+    }
+
+    if (params == NULL) {
+        return RET_OSSL_OK;
+    }
+
+    switch (ctx->type) {
+    case CKK_EC:
+        p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_GROUP_NAME);
+        if (p) {
+            int i;
+            if (p->data_type != OSSL_PARAM_UTF8_STRING) {
+                return RET_OSSL_ERR;
+            }
+            for (i = 0; ec_name_to_params[i].name != NULL; i++) {
+                if (strcmp(ec_name_to_params[i].name, p->data) == 0) {
+                    break;
+                }
+            }
+            if (ec_name_to_params[i].name == NULL) {
+                P11PROV_raise(ctx->provctx, CKR_ARGUMENTS_BAD,
+                              "Unknown Curve %*s", (int)p->data_size,
+                              (char *)p->data);
+                return RET_OSSL_ERR;
+            }
+            ctx->data.ec.ec_params = ec_name_to_params[i].ec_param;
+            ctx->data.ec.ec_params_size = ec_name_to_params[i].ec_param_size;
+        }
+        break;
+    case CKK_EC_EDWARDS:
+        p = OSSL_PARAM_locate_const(params, "p11prov_edname");
```

Comment:
Should we make some constant for this string, like `P11PROV_EDNAME` to avoid keeping constant strings all over the place?